daemons/based/pacemaker-based.c

```diff
@@ -134,11 +134,11 @@ based_metadata(void)
     const char *desc_long = "Cluster options used by Pacemaker's Cluster "
                             "Information Base manager";
 
-    gchar *s = pcmk__cluster_option_metadata(name, desc_short, desc_long,
-                                             pcmk__opt_context_based);
+    char *s = pcmk__cluster_option_metadata(name, desc_short, desc_long,
+                                            pcmk__opt_context_based);
 
     printf("%s", s);
```

Comment:
All of the daemons should be using output_xml() (like pacemaker-schedulerd) instead of printf()

Reply:
That's coming, I just postponed it to later.

https://github.com/ClusterLabs/pacemaker/pull/3338/commits/1a8c6e2f

Reply:
Ah cool. However we don't need to create an output object, all the daemons already have one for their --version argument etc. We don't want to force XML output; it's a bit confusing, but any output format can display XML via output_xml(): text outputs it raw (which is important for pcs) while xml wraps it in CDATA so it doesn't break the API schema.

Reply:
It's done that way for backward compatibility. Currently, a daemon outputs raw XML for "text" output and CDATA for "xml" output.

As of https://github.com/ClusterLabs/pacemaker/commit/1a8c6e2f, `pcmk__cluster_option_metadata()` accepts an "xml" output object and uses it to output raw XML. We want the daemon output to stay the same (for example if the daemon uses "text" output). So we create a temporary XML output object that doesn't print anything, make a copy of its result, and call `output_xml()` on the result using the daemon's main output object. It's inefficient, but these are one-shot calls.

Later, we turn `pcmk__cluster_option_metadata()` into an output message function and then add a default one for "text": https://github.com/ClusterLabs/pacemaker/pull/3338/commits/e363cee1. The "text" one is a human-readable list, as we discussed previously.

Reply:
Huh. What I said above is MOSTLY correct. However, I thought in that future commit, we called `output_xml()` on the resulting `PCMK_XE_RESOURCE_AGENT` XML object. Instead we still call `print()` for the controller, CIB manager, and fencer.

Technically that's still the most backward-compatible way to do it, since those two daemons have always printed raw XML to stdout even if the main output object is xml-format... it would probably be fine to call `output_xml` on it though (and wouldn't be any more complex).

Reply:
I think I get it now ... for the new command, we want the text output to be friendly, so that's what your custom message will do (not output_xml()). For the daemons to preserve their behavior, they'll need to call output_xml() directly instead of the custom message. That will preserve the current behavior for text and none, and output XML in an API-standard way. (There's no schema for daemon XML output, so we don't have to worry about that.)

Reply:
Right :)